### DISCUSSION TITLE
Corne 42 custom pinkie buttons per layer, ZMK

### DISCUSSION THREAD

**Discussion by dpeley** [manna-harbour/miryoku#304] [Q&A] [answered]
I have been searching for over a week to try and find this answer, so I'm sorry if it's somewhere obvious. I'm trying to create a custom mapping for the Miryoku ZMK custom_config file to let me map the extra pinkie keys on the Corne 6 column (42 key) differently **per layer**. Can anyone point me in the right direction for that?

**Comment by manna-harbour** [comment, marked as answer]
See the _To change the mapping per layer_ paragraph in the _custom_config.h / Mapping / Miryoku ZMK_ section at https://github.com/manna-harbour/miryoku/discussions/85. Also see the _For complete examples, see example custom_config.h shared by members of the community_ link there.

E.g to map the left hand top row outer pinkie column key on the Tap layer to Escape:
```
#if defined (MIRYOKU_KEYBOARD_CORNE)

#define XXX &none

#define MIRYOKU_LAYERMAPPING_TAP( \
     K00, K01, K02, K03, K04,      K05, K06, K07, K08, K09, \
     K10, K11, K12, K13, K14,      K15, K16, K17, K18, K19, \
     K20, K21, K22, K23, K24,      K25, K26, K27, K28, K29, \
     N30, N31, K32, K33, K34,      K35, K36, K37, N38, N39 \
) \
&kp ESC K00  K01  K02  K03  K04       K05  K06  K07  K08  K09  XXX \
XXX  K10  K11  K12  K13  K14       K15  K16  K17  K18  K19  XXX \
XXX  K20  K21  K22  K23  K24       K25  K26  K27  K28  K29  XXX \
               K32  K33  K34       K35  K36  K37

#endif
```

Here is another example of a custom per layer mapping: https://github.com/manna-harbour/miryoku/discussions/158.

**Reply by dpeley** [reply]
Thank you so much for the reply! I've gone and tried that, but it doesn't seem to be working, here's my current [config](https://github.com/dpeley/DanCrkbd5ZmkMiryoku/blob/28cc7b1ea7e877d9b68db696e8490212603a534c/miryoku/custom_config.h)

I know that I don't have to include the layers that I'm not going to change, but I'm intending on tweaking most of them, so I threw them all down, and only put in a few small changes to start. The build seems to work with no errors, but after flashing, all I have is the unchanged layout with none of the pinkies assigned.

When I build, I'm using the "nice_nano_v2", and "corne_left,corne_right" options.

**Reply by manna-harbour** [reply]
That looks ok.

You can verify that the file is being used by adding `#error` somewhere inside the `if` / `endif` block.

Also check that you are flashing the files downloaded from the workflow build, as it's easy to accidentally flash the wrong file, and that the board resets itself after flashing.

**Reply by dpeley** [reply]
I've been deleting all the files after trying, so no danger of using the wrong one, yes the board resets afterwards, and when I tried building again with the #error, it stated this:

`/__w/DanCrkbd5ZmkMiryoku/DanCrkbd5ZmkMiryoku/miryoku_zmk/config/../miryoku/custom_config.h:6:2: error: #error 
      6 | #error
        |  ^~~~~
  CMake Error at /__w/DanCrkbd5ZmkMiryoku/DanCrkbd5ZmkMiryoku/zmk/zephyr/cmake/modules/extensions.cmake:3885 (message):
    failed to preprocess devicetree files (error code 1):`
    
    Thank you for your patience and all your hard work on this wonderful layout

**Reply by dpeley** [reply]
Actually, now that I think about it, I know for sure that this first build was used and successfully flashed, cause the key map on the corne was previously qwerty before flashing, and the one that I built with the custom_config file was the default (colemakDH). The Colemak was successfully mapped, but the pinkies weren't.

**Reply by manna-harbour** [reply]
Sorry, my example was bad. It should have been MIRYOKU_**LAYER**MAPPING_TAP! I'll correct it in the original comment.

**Reply by dpeley** [reply]
This did it! Thank you so much!
